### PR TITLE
Mod. to helm-allow-mouse (and doc. improvements) to allow global mouse bindings

### PR DESCRIPTION
**Pull request by rswgnu**
First, thanks for the great work on helm.

Sometimes one wants to have consistent mouse bindings across many different modes where the global mouse settings determine how to handle the specific context of each mouse click.  This pull request allows for this behavior via a flag setting prior to helm initialization.

This change does not affect helm behavior unless a user explicitly sets helm-allow-mouse to 'no-mode-bindings, in which case all local and text property helm mouse bindings are skipped and never activated.  That is all it does so it should be easy to integrate without affecting any other uses while allowing for more flexible mouse key binding.

One specific use case is with my package, Hyperbole, which provides two context-specific global mouse keys.  If this change is made, the next release of Hyperbole will include helm compatibility and will allow for random access to items in helm completion buffers even when helm is not active (if the completion buffer is still available).  If helm is active, after each click, point is returned to the minibuffer so that mouse usage does not impede keyboard usage.

### CODE REVIEW

**Comment by thierryvolpiatto** [comment]
Thanks for your contribution, I have added some comments.
Also not sure what you want to do when switching to a helm-buffer not activated.

**Comment by rswgnu** [comment]
Thanks, Thierry.  Where would I find your comments?  I don't see them here or anywhere in helm.el on the master branch.  Are you planning to merge this?

For one usage example, see: https://www.gnu.org/software/hyperbole/man/hyperbole.html#Smart-Keys

Hyperbole globally binds 2 mouse keys that do all sorts of control in different modes so that behavior is more consistent across all these modes.  So in helm-major-mode, it will allow random access to helm completions as well as showing you the actions of a helm completion when its help function (Smart Assist Key) bound to a mouse key/button is used.

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> Thanks, Thierry. Where would I find your comments?

You should now see requested changes to merge this PR.

> I don't see them here or anywhere in helm.el on the master branch.

Before finishing review they are in the diff of your changes on github.

> Are you planning to merge this?

Not until the requested changes are applied, for now your PR is breaking
helm.

> For one usage example, see: https://www.gnu.org/software/hyperbole/man/hyperbole.html#Smart-Keys
>
> Hyperbole globally binds 2 mouse keys that do all sorts of control in
> different modes so that behavior is more consistent across all these
> modes. So in helm-major-mode, it will allow random access to helm
> completions as well as showing you the actions of a helm completion
> when its help function (Smart Assist Key) bound to a mouse key/button
> is used.

Normally helm-buffer's should not be used (accessed) without activating helm, use
helm-resume for this.

--
Thierry

File: helm.el [hunk 1]
```diff
@@ -4142,7 +4155,10 @@ If PRESERVE-SAVED-ACTION is non-`nil', then save the action."
                          (and (assq 'accept-empty source) ""))
                      source))
     (unless preserve-saved-action (setq helm-saved-action nil))
-    (when (and selection action) (funcall action selection))))
+    ;; Selection may be a string when `helm-allow-mouse' is non-nil when
+    ;; a section header is selected; make sure we don't try to activate
+    ;; such selections with the (listp) test below.
+    (when (and selection (listp selection) action) (funcall action selection))))
```
[the remark this answers is not in the document]
Comment (rswgnu) [reply helm.el]: Yes, I saw this today and am working on a fix.  There is a simple change there but I am also adding a fix to the helm-allow-mouse specification so when it is changed, if local mouse bindings are not desired, they are disabled.  Please wait for the update.

**Comment by rswgnu** [comment]
I have finished the fixes requested and have done some basic testing on them.  I will make a new pull request for the latest commit.

File: helm.el [hunk 2]
```diff
@@ -559,14 +559,27 @@ source name in this variable."
   :type 'boolean)
 
 (defcustom helm-allow-mouse nil
-  "Prevent mouse usage during the helm session when non-nil.
+  "Allow mouse usage during the helm session when non-nil.
 
-Note that this also allow moving out of minibuffer when clicking
-outside of `helm-buffer', up to you to get back to helm by clicking
-back in `helm-buffer' of minibuffer."
+If set to 'no-mode-bindings, no local nor text-specific mouse bindings
```
[the remark this answers is not in the document]
Comment (rswgnu) [reply helm.el]: This has been done in the latest commit.

File: helm.el (hunk 1 above)
[the remark this answers is not in the document]
Comment (rswgnu) [reply helm.el]: This is fixed in the latest commit.

**Comment by thierryvolpiatto** [comment]
I am sorry, but now mouse is no more working in helm.
It seems you make a lot of efforts making mouse working in dead helm buffers, why ?

**Comment by rswgnu** [comment]
Sorry, yes there is an inverted flag test.  I will work on a fix.

**Comment by rswgnu** [comment]
All I really want to be able to do when helm is not active (but after a helm completion buffer has been generated and left around) is to be able to mark selections with a mouse button in that buffer and then execute the marked selection's default action with a mouse button, without having the minibuffer active or helm controlling things.

We have these 3 functions: 	(helm-mark-current-line), (helm-get-selection), and (helm-execute-persistent-action) but they generally insist on having an active helm session to work properly.

Is there any straightforward way to get this behavior in two mouse button functions or is there some limit to the helm design that requires it to be active in minibuffer in order to process actions?  Understanding this would help a lot.

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> All I really want to be able to do when helm is not active (but after
> a helm completion buffer has been generated and left around) is to be
> able to mark selections with a mouse button in that buffer and then
> execute the marked selection's default action with a mouse button,
> without having the minibuffer active or helm controlling things.

I don't believe this is an interesting feature for somebody used to
helm.  I think you want this feature only because you are used to emacs
standard use of buffers.
Using M-x helm-resume allows you to do all what you want and more (i.e
continue searching).

> We have these 3 functions: (helm-mark-current-line),
> (helm-get-selection), and (helm-execute-persistent-action) but they
> generally insist on having an active helm session to work properly.
>
> Is there any straightforward way to get this behavior in two mouse
> button functions or is there some limit to the helm design that
> requires it to be active in minibuffer in order to process actions?

You can, see how helm-grep allow saving its buffers, but again I don't
want to provide such a feature for all helm buffers, this will introduce
a big confusion about the helm usage and also many bugs.

--
Thierry

**Comment by rswgnu** [comment]
I simplified the changes greatly (use the All option to view the changes all together).  I got things working properly so you can instantly switch between global, local and no mouse key bindings in helm.  This allows you to test out and utilize different mouse handling techniques without affecting the standard helm mouse bindings.  I have already found it useful to utilize global mouse bindings I am used to.

I only tested with a single helm-info buffer but wrote it to handle multiple buffers as helm does.  Try it out with more complex scenarios and let me know if it works properly.  To switch mouse modes, use each of the following lines, one at a time:

(custom-set-variables '(helm-allow-mouse 'global-mouse-bindings))
(custom-set-variables '(helm-allow-mouse t))
(custom-set-variables '(helm-allow-mouse nil))

**Comment by thierryvolpiatto** [comment]
I have an error at first try, see backtrace below.
Please remove all unrelated changes to mouse in your PR so that the diff can be read easily.

Debugger entered--Lisp error: (wrong-type-argument window-valid-p #<window 12>)
  window-minibuffer-p(#<window 12>)
  mouse-minibuffer-check((drag-mouse-2 (#<window 12> 771 (25 . 215) 82426871 nil 771 (3 . 12) nil (1 . 11) (8 . 17)) (#<window 3 on helm.el> 22699 (25 . 579) 82427039 nil 22699 (3 . 34) nil (1 . 1) (8 . 17))))
  mouse-set-point((drag-mouse-2 (#<window 12> 771 (25 . 215) 82426871 nil 771 (3 . 12) nil (1 . 11) (8 . 17)) (#<window 3 on helm.el> 22699 (25 . 579) 82427039 nil 22699 (3 . 34) nil (1 . 1) (8 . 17))))
  mouse-yank-primary((drag-mouse-2 (#<window 12> 771 (25 . 215) 82426871 nil 771 (3 . 12) nil (1 . 11) (8 . 17)) (#<window 3 on helm.el> 22699 (25 . 579) 82427039 nil 22699 (3 . 34) nil (1 . 1) (8 . 17))))
  call-interactively(mouse-yank-primary nil nil)
  command-execute(mouse-yank-primary)

**Comment by rswgnu** [comment]
I am testing mouse-yank-primary on button2 and I am not getting this error,
whether or not helm is active.
Could you tell me what if anything you have configured and then what
operations you are doing and approximately where?
What did you set helm-allow-mouse to, if anything?
I assume you have all 4 patches loaded.

----

Separately, I am not sure how to separate out the comment changes.  They
are pretty short (one or two liners) and easy to see, everything else is a
function change and the --remap minor mode was renamed to --disable to
better describe its function.

Bob

**Comment by thierryvolpiatto** [comment]
helm-allow-mouse is set to t (but not from custom), then I hit mouse-1 to select a candidate and after mouse-2 and I have an error.
If users have to go to custom to modify their helm-allow-mouse settings, we will have tons of bugreports.

**Comment by thierryvolpiatto** [comment]
Also please remove helm-info.el modification which is unrelated to this PR, I will make a commit on master to fix this require.

Thanks.

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 1:24 AM, Thierry Volpiatto <notifications@github.com
> wrote:

> helm-allow-mouse is set to t (but not from custom), then I hit mouse-1 to
> select a candidate and after mouse-2 and I have an error.
>
​Ok, after following your setup, things initially worked fine but after
about 6 good selections, the 7th triggered the same error you received.
Somehow a global mouse binding is leaking through when you think you are in
local mouse mode.  I'll have to look at it.

Unfortunately, setq does not run the needed :set function from a custom
option and so does not initialize the option fully.  In this case, it won't
switch the text property mouse bindings between local and global, so unless
setq were ovridden, I don't see a way around this.  Personally, I think
setq should do the right thing for options but Emacs has not yet evolved
that way.

Bob

If users have to go to custom to modify their helm-allow-mouse settings, we
> will have tons of bugreports.
>
​So you want me to commit another change with that one change backed out?

Bob
​

**Comment by thierryvolpiatto** [comment]
After some more tests I see other things like helm-occur are broken with your changes, briefly reading the code I also saw you are using lexical-binding which is wrong.
Use of mouse-set-point is actually triggering the error.

If what you want is only having usage of mouse like anywhere in emacs from helm, this can be done with very few changes, your actual patch is quite big and it is hard to find out what's wrong.

**Comment by xiongtx** [comment]
> Personally, I think `setq` should do the right thing for options but Emacs has not yet evolved
that way.

How about `customize-set-value`?

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> ​So you want me to commit another change with that one change backed out?

No, actually your patch involves too much changes which I guess will
introduce many bugs (in addition of the ones I notified in previous
post), please start back from master with a simple `(when (eq helm-allow-mouse t)
[...])` where needed (in 4 or 5 places, not sure exactly) and let me
know what you are missing when doing `(setq helm-allow-mouse 'global)`,
then we could add progressively features from there with little commits
allowing to track down any errors.

Thanks.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Tianxiang Xiong <notifications@github.com> writes:

>     Personally, I think setq should do the right thing for options but Emacs has not yet evolved
>     that way.
>    
> How about customize-set-value?

The problem is to be backward compatible with previous user configs.

-- 
Thierry

**Comment by rswgnu** [comment]
Thanks for the feedback and pointers.

On Fri, Sep 15, 2017 at 4:29 AM, Thierry Volpiatto <notifications@github.com
> wrote:

>
> > How about customize-set-value?
>
> The problem is to be backward compatible with previous user configs.


​So is your suggestion to see if we can avoid custom entirely?  It is used
just to force a helm-refresh after the value changes (but deferred until
helm is invoked in each helm buffer).  If we could add the tracking of
helm-allow-mouse changes directly to helm initialization, then the
customization wouldn't be necessary.

I will work on a constraining the patch more and the fixes.  I am new to
helm as you can tell but find it useful as I am able to integrate it with
my other packages.

Bob

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> I will work on a constraining the patch more and the fixes.

You should start as suggested in previous post and make small commits,
not one huge patch which is hard to bisect.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> ​So is your suggestion to see if we can avoid custom entirely?

No, only its :set function.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Thierry Volpiatto <thievol05@zoho.eu> writes:

> Robert Weiner <notifications@github.com> writes:
>
>> I will work on a constraining the patch more and the fixes.
>
> You should start as suggested in previous post and make small commits,
> not one huge patch which is hard to bisect.

I created a new branch called global_mouse where you can start from.

-- 
Thierry

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 2:40 AM, Thierry Volpiatto <notifications@github.com
> wrote:

> After some more tests I see other things like helm-occur are broken with
> your changes
>



> , briefly reading the code I also saw you are using lexical-binding which
> is wrong.
>
​That was related to the use of customization, which we seem to need to
remove anyway.​

> Use of mouse-set-point is actually triggering the error.
>
​mouse-set-point was used to force point to the position of the mouse
click; otherwise, marking of the selection did not work properly; any
suggestions?​

> If what you want is only having usage of mouse like anywhere in emacs from
> helm, this can be done with very few changes
>
​Yes, but also including selection and activation of helm items.

Bob
​

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 9:46 AM, Thierry Volpiatto <notifications@github.com
> wrote:

>
> Thierry Volpiatto <thievol05@zoho.eu> writes:
>
> > Robert Weiner <notifications@github.com> writes:
> >
> >> I will work on a constraining the patch more and the fixes.
> >
> > You should start as suggested in previous post and make small commits,
> > not one huge patch which is hard to bisect.
>
> I created a new branch called global_mouse where you can start from.


​Ok, thanks.

Bob
​

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

>> , briefly reading the code I also saw you are using lexical-binding which
>> is wrong.
>>
> ​That was related to the use of customization, which we seem to need to
> remove anyway.​

Don't use lexical-binding which is part of cl.el, never, helm already use
lexical code.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> ​Yes, but also including selection and activation of helm items.

I want to clearly understand this, really.

Actually we can: 
- click on a candidate to select it
- click on this candidate with mouse-2 to run default action on it
- click anywhere else and do what you do usually with mouse (selection
etc...)

Can you describe exactly what is missing with actual helm when
helm-allow-mouse is enabled ?

It seems it is related to your hyperbole package but it stills really
opaque what you want to achieve.

-- 
Thierry

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 10:53 AM, Thierry Volpiatto <
notifications@github.com> wrote:

>
> Robert Weiner <notifications@github.com> writes:
>
> > ​Yes, but also including selection and activation of helm items.
>
> I want to clearly understand this, really.
>
> Actually we can:
> - click on a candidate to select it
> - click on this candidate with mouse-2 to run default action on it
> - click anywhere else and do what you do usually with mouse (selection
> etc...)
>
> Can you describe exactly what is missing with actual helm when
> helm-allow-mouse is enabled ?
>

​Helm's mouse bindings attached to text properties within the match buffer
do not allow for mouse key rebinding and they override global mouse key
bindings that a user may have and may want to use.

In the specific case of GNU Hyperbole, ​​Hyperbole uses global mouse
bindings on either mouse-2 and mouse-3​ (or shift mouse-2 and shift
mouse-3) to do dozens of things in a consistent and mode-specific manner.
It takes a few lines of code to interface Hyperbole to new modes
typically.  Hyperbole utilizes both the press down event and the release
event on each mouse key it uses.  Helm's existing code is not compatible
with this use.  Some things Hyperbole or other packages might do with Helm
if all global mouse keys were available and not shadowed:

    1. Point at a helm completion and show its list of actions with the
mouse
    2. Use the helm completion buffer as a menu and perform the action for
one match item after another {like C-j does}
    3. Select a helm completion buffer and move it to another window or
frame
    4. Filter the match list further by selecting terms visible within the
match list

Basically, it expands the possible UI to a more flexible mouse-based
approach that people can experiment with, without knowing much about Helm's
internals.

I hope that helps.

Bob

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> On Fri, Sep 15, 2017 at 10:53 AM, Thierry Volpiatto <
> notifications@github.com> wrote:
>
>>
>> Robert Weiner <notifications@github.com> writes:
>>
>> > ​Yes, but also including selection and activation of helm items.
>>
>> I want to clearly understand this, really.
>>
>> Actually we can:
>> - click on a candidate to select it
>> - click on this candidate with mouse-2 to run default action on it
>> - click anywhere else and do what you do usually with mouse (selection
>> etc...)
>>
>> Can you describe exactly what is missing with actual helm when
>> helm-allow-mouse is enabled ?
>>
>
> ​Helm's mouse bindings attached to text properties within the match buffer
> do not allow for mouse key rebinding and they override global mouse key
> bindings that a user may have and may want to use.
>
> In the specific case of GNU Hyperbole, ​​Hyperbole uses global mouse
> bindings on either mouse-2 and mouse-3​ (or shift mouse-2 and shift
> mouse-3) to do dozens of things in a consistent and mode-specific manner.
> It takes a few lines of code to interface Hyperbole to new modes
> typically. Hyperbole utilizes both the press down event and the release
> event on each mouse key it uses. Helm's existing code is not compatible
> with this use. Some things Hyperbole or other packages might do with Helm
> if all global mouse keys were available and not shadowed:
>
> 1. Point at a helm completion and show its list of actions with the
> mouse

mouse-3 in an active helm session is already doing this, as I said I
don't want to develop a workflow that use dead helm buffers.

> 2. Use the helm completion buffer as a menu and perform the action for
> one match item after another {like C-j does}

Same.

> 3. Select a helm completion buffer and move it to another window or
> frame

Same.

> 4. Filter the match list further by selecting terms visible within the
> match list

Same.

> Basically, it expands the possible UI to a more flexible mouse-based
> approach that people can experiment with, without knowing much about Helm's
> internals.

No, this allow only to reuse the dead helm buffers.

> I hope that helps.

Thanks for explanation, but AFAIU what you want to provide with a global
setting of helm-allow-mouse is a way to use dead helm buffers.
I spent lot of time to implement separated keymaps in helm and also
recently mouse support, I don't want to break this mechanism and have to
do a rewrite of all this just to implement a workflow which I believe
is inefficient compared to the helm way to search and narrow buffers.
  
-- 
Thierry

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 2:36 PM, Thierry Volpiatto <notifications@github.com
> wrote:

>
> Thanks for explanation, but AFAIU what you want to provide with a global
> setting of helm-allow-mouse is a way to use dead helm buffers.
>

​I want to use global mouse keys in helm buffers both when they are alive
and not alive.  If not alive, the first mouse click could be required to
use helm-resume (which is what I am doing at present in testing) before
invoking any other helm actions.  So aside from resuming helm, you could
assume that the usage is when helm is active.

I want to be able to bind custom mouse keys for use in helm and not have
the help-echo strings show up and display improper documentation for what
the keys in use do.

[mouse-3] in helm often shows a menu of actions, it doesn't do what C-j
does, which is invoke the persistent action for the selected item without
exiting the minibuffer.  How can you do this in helm now with the mouse?
​

> I spent lot of time to implement separated keymaps in helm and also
> recently mouse support, I don't want to break this mechanism


​Certainly.
​

> and have to
> do a rewrite of all this just to implement a workflow which I believe
> is inefficient compared to the helm way to search and narrow buffers.


I am new to helm, so that is a bit of a barrier but I think we are getting
through it.​
I have made the incremental commits you wanted in the new branch and will
push it soon.
Even though it is not completely finished, I will push it so you can have a
look.

The issue in helm-occur is that there were local key bindings set in
helm-regexp.el
and helm-grep.el that I hadn't dealt with yet but will on this new branch.

If you could suggest how to avoid the addition of the mouse-set-point call
to move
the selection overlay, that would help me resolve it.

Bob

**Comment by thierryvolpiatto** [comment]
Robert Weiner <notifications@github.com> writes:

> On Fri, Sep 15, 2017 at 2:36 PM, Thierry Volpiatto <notifications@github.com
>> wrote:
>
>>
>> Thanks for explanation, but AFAIU what you want to provide with a global
>> setting of helm-allow-mouse is a way to use dead helm buffers.
>>
>
> ​I want to use global mouse keys in helm buffers both when they are alive
> and not alive. If not alive, the first mouse click could be required to
> use helm-resume (which is what I am doing at present in testing) before
> invoking any other helm actions. So aside from resuming helm, you could
> assume that the usage is when helm is active.

I am not sure your approach is good, it looks to me just like wanting to
animate a html buffer with the browser, why not just accessing this html
file with the browser directly ?

For example bookmark provide a way to do this with handlers, you do not
access the .bmk file directly to activate for example a w3m bookmark.

I guess that for helm it is the same (I already made a bookmark handler
for helm-find-files) you could save a link from a helm session, put this
link somewhere (in hyperbole or anywhere) and click on this link to
switch back in the helm-session with helm-resume, but accessing first
the raw helm buffer and try to reanimate it look strange just like
opening a html file in a text editor and then activate the browser to
read this file.

This is just my feeling.

> I want to be able to bind custom mouse keys for use in helm and not have
> the help-echo strings show up and display improper documentation for what
> the keys in use do.
>
> [mouse-3] in helm often shows a menu of actions, it doesn't do what C-j
> does, which is invoke the persistent action for the selected item without
> exiting the minibuffer. How can you do this in helm now with the mouse?

You can't, I wanted to have only a very simple implementation of mouse
in helm to do the very minimum: Select, execute and popup the menu
allowing to click on diverse actions.
 ​
> Even though it is not completely finished, I will push it so you can have a
> look.

Good thanks, I will have a look.

> The issue in helm-occur is that there were local key bindings set in
> helm-regexp.el
> and helm-grep.el that I hadn't dealt with yet but will on this new branch.

I have also noticed that the helm-sources-using-default-as-input
mechanism was broken.

> If you could suggest how to avoid the addition of the mouse-set-point call
> to move
> the selection overlay, that would help me resolve it.

Don't know yet, I have to look carefully at the code as soon as I have
time.

-- 
Thierry

**Comment by rswgnu** [comment]
On Fri, Sep 15, 2017 at 4:01 PM, Thierry Volpiatto <notifications@github.com
> wrote:

>
> For example bookmark provide a way to do this with handlers, you do not
> access the .bmk file directly to activate for example a w3m bookmark.
>
> I guess that for helm it is the same (I already made a bookmark handler
> for helm-find-files) you could save a link from a helm session, put this
> link somewhere (in hyperbole or anywhere) and click on this link to
> switch back in the helm-session with helm-resume, but accessing first
> the raw helm buffer and try to reanimate it look strange just like
> opening a html file in a text editor and then activate the browser to
> read this file.
>

​Fair enough.  I am open to some approach like this, though
I don't see much of a difference since the same information
could be stored with the inactive buffer and used by helm
upon resume.  I guess with bookmarks you could force a single
helm match buffer into different states with different bookmarks.

I am mainly interested in navigating the different helm subsystems
(grep, ff, etc.) via menu and then restoring inactive sessions for
each subsystem when selecting them and then selecting matches or
getting help for matches, all from the mouse.

I just tried the find-file bookmark but it did not
restore the match set from when the bookmark was stored, which is
what we'd want to operate on with the mouse upon restoration and
is why I was drawn to the inactive helm buffer originally.

> I want to be able to bind custom mouse keys for use in helm and not have
> > the help-echo strings show up and display improper documentation for what
> > the keys in use do.
> >
> > [mouse-3] in helm often shows a menu of actions, it doesn't do what C-j
> > does, which is invoke the persistent action for the selected item without
> > exiting the minibuffer. How can you do this in helm now with the mouse?
>
> You can't, I wanted to have only a very simple implementation of mouse
> in helm to do the very minimum
> ​.
>

​That is why I'd like it to be extendible.  Your mouse use is
minimal but others may not be.  Additionally, most functions
the Hyperbole mouse keys can do, so can its keyboard bindings.
I am already making use of this fact even before this code
is finished.

> ​
>
> ​​
> > The issue in helm-occur is that there were local key bindings set in
> ​​
> > helm-regexp.el
> ​​
> > and helm-grep.el that I hadn't dealt with yet but will on this new
> branch.
>

​​
>
> ​​
> I have also noticed that the helm-sources-using-default-as-input
> ​​
> mechanism was broken.


​Can you give an example of how to trigger this error?​

>
>
> > If you could suggest how to avoid the addition of the mouse-set-point
> call
> > to move
> > the selection overlay, that would help me resolve it.
>
> Don't know yet, I have to look carefully at the code as soon as I have
> time.


​Okay.

Thanks,

Bob
​

**Comment by thierryvolpiatto** [comment]
I close this PR, let continue in #1865
